controlplane/kubeadm/internal/controllers/controller.go

```diff
@@ -910,6 +907,43 @@ func (r *KubeadmControlPlaneReconciler) reconcileControlPlaneConditions(ctx cont
 	return nil
 }
 
+func reconcileMachineUpToDateCondition(_ context.Context, controlPlane *internal.ControlPlane) {
+	machinesNeedingRollout, _, rolloutConditionsReasons := controlPlane.MachinesNeedingRollout()
+	machinesNeedingRolloutNames := sets.New(machinesNeedingRollout.Names()...)
+
+	for _, machine := range controlPlane.Machines {
+		if !machine.DeletionTimestamp.IsZero() {
+			v1beta2conditions.Set(machine, metav1.Condition{
+				Type:    clusterv1.MachineUpToDateV1Beta2Condition,
+				Status:  metav1.ConditionFalse,
+				Reason:  clusterv1.MachineNotUpToDateV1Beta2Reason,
+				Message: "Machine DeletionTimestamp is set",
+			})
+			continue
+		}
+		if machinesNeedingRolloutNames.Has(machine.Name) {
+			message := ""
+			if reasons, ok := rolloutConditionsReasons[machine.Name]; ok {
+				message = strings.Join(reasons, "; ")
+			}
+
+			v1beta2conditions.Set(machine, metav1.Condition{
+				Type:    clusterv1.MachineUpToDateV1Beta2Condition,
+				Status:  metav1.ConditionFalse,
```

Comment:
Wondering if Unknown would be more apprioriate or alternatively not handling the deletionTimestamp and calculating if this Machine is up-to-date also for deleting Machines (my preferred option if possible).
I think we only have to additionally calculate the reasons for all Machines, we don't have to add Machines with deletionTimestamp to machinesNeedingRollout


I think it's not correct to assume this Machine is not up-to-date just because it gets deleted (e.g. if someone deletes the Machine manually, remediation, if the entire KCP gets deleted). Seems a bit strange to me that Machines flip to UpToDate false as soon as they get deleted. Also this Means that as soon as Machine actually gets rolled out the reason for the rollout gets lost and overwritten.

Reply:
fixed